### PR TITLE
Fuzzing: Add 3 fuzzers

### PR DESCRIPTION
**Pull request by AdamKorcz**
This PR adds 3 fuzzers for Distribution and a build script to run the fuzzers continuously through OSS-fuzz.

A PR has also been set up on the OSS-fuzz side to integrate Distribution: https://github.com/google/oss-fuzz/pull/6014. Two of the fuzzers that were committed on the OSS-fuzz side are moved upstream in this PR and a third fuzzer is added.

Once this PR gets merged, I will amend the changes on OSS-fuzz to invoke the build script in this PR.

For those unfamiliar with fuzzing and/or OSS-fuzz: Fuzzing (or fuzz testing) is a way of testing software whereby pseudo-random data is passed to a target application with the goal of finding bugs and vulnerabilities. It has been useful in finding bugs, including security-critical bugs, in many open source Go projects to date. Some can be found here, https://github.com/dvyukov/go-fuzz#trophies and others include [this CVE](https://groups.google.com/g/kubernetes-announce/c/ALL9s73E5ck/m/4yHe8J-PBAAJ) in Kubernetes and [this CVE](https://groups.google.com/g/golang-announce/c/NpBGTTmKzpM/m/fLguyiM2CAAJ) in the Go standard library found by fuzzing Go-ethereum.

OSS-fuzz is a free service offered by Google into which critical open source projects can integrate and have their fuzzers run continuously. If any bugs are found, maintainers will get notified with an email containing a link to a detailed bug reports.
The fuzzers in this PR are all implemented by way of [go-fuzz](https://github.com/dvyukov/go-fuzz) which is the Golang engine supported by OSS-fuzz.
To complete the OSS-fuzz PR, at least one maintainers email address is needed for bug reports.

Other projects integrated into OSS-fuzz include Kubernetes, runC and containerd.

### CODE REVIEW

File: reference/fuzz.go
Comment (brackendawson) [inline reference/fuzz.go:5]: As `reference` is a package intended to be imported by others, I would prefer not to add `FuzzParseNormalizedNamed` as an exported function. Can you move this fuzz to its own package under an `internal` directory?

**Review by brackendawson — CHANGES_REQUESTED** [review]

File: reference/fuzz.go
Comment (AdamKorcz) [reply reference/fuzz.go]: I didn't find any existing `internal` directory, so instead of creating one for a single fuzzer, perhaps we could unexport the fuzzer, and when building it through OSS-fuzz we export it. This will work fine from the perspective of building it successfully through OSS-fuzz.
I did mention earlier that the fuzz tests need to be exported. This is only true when we build them, and they can easily remain unexported in their respective packages when we are not building them.

The current state of the files in the PR reflect the above suggestion. Let me know what you think.

Comment (brackendawson) [reply reference/fuzz.go]: If that works then that's even nicer, thanks. Could we unexport the others?

Comment (AdamKorcz) [reply reference/fuzz.go]: Done!

Comment (AdamKorcz) [reply reference/fuzz.go]: @brackendawson Do you have a suggestion regarding the current failing CI test? The fuzzers are now not invoked.

```
reference/fuzz.go:8:6: `fuzzParseNormalizedNamed` is unused (deadcode)
func fuzzParseNormalizedNamed(data []byte) int {
     ^
configuration/fuzz.go:10:6: `parserFuzzer` is unused (deadcode)
func parserFuzzer(data []byte) int {
     ^
registry/api/v2/fuzz.go:7:6: `fuzzParseForwardedHeader` is unused (deadcode)
func fuzzParseForwardedHeader(data []byte) int {
```

Comment (brackendawson) [reply reference/fuzz.go]: Oh jeez, very much looking forward to native fuzzing. For now just `// nolint;deadcode` them (https://golangci-lint.run/usage/false-positives/#nolint).

**Comment by AdamKorcz** [comment]
The 3 fuzzers  in this PR along with a few more have been merged on CNCF's fuzzing repo: https://github.com/cncf/cncf-fuzzing/tree/main/projects/distribution

Can we get a maintainers email address for https://github.com/google/oss-fuzz/pull/6014? It is all ready with the fuzzers from the cncf-fuzzing repo.

**Comment by milosgajdos** [comment]
You will need to rebase to re-kick the actions. There is no way I can kick the GHA run otherwise 🤔

**Review by milosgajdos — APPROVED** [review]

**Comment by AdamKorcz** [comment]
Great if we can finally get this merged @milosgajdos ! Could you see my initial message where I describe the OSS-Fuzz integration: are there any emails I can put in to receive bug reports? 

Notice that there are quite a few more fuzzers ready for Distribution and we would like to get these merged in as well https://github.com/cncf/cncf-fuzzing/tree/main/projects/distribution as such, how do you want to go about this? I can create another PR that uploads all of the fuzzers in one go?

**Comment by milosgajdos** [comment]
> Great if we can finally get this merged @milosgajdos ! Could you see my initial message where I describe the OSS-Fuzz integration: are there any emails I can put in to receive bug reports?

What do you mean by "bug reports"? Security vulnerabilities found by the fuzzers?

> Notice that there are quite a few more fuzzers ready for Distribution and we would like to get these merged in as well https://github.com/cncf/cncf-fuzzing/tree/main/projects/distribution as such, how do you want to go about this? I can create another PR that uploads all of the fuzzers in one go?

I'd welcome adding more fuzzers, indeed. Personally, I'd have no issue with having a single PR covering all of those, but I'm but a single maintainer. I can bring it up with other folks at the next call. In the meantime, can you please rebase so the GHA are run?

**Comment by DavidKorczynski** [comment]
> 
> What do you mean by "bug reports"? Security vulnerabilities found by the fuzzers?
> 

The goal is to find security and also reliability issues, so the fuzzer will report on both. There is most likely some false positives too if a fuzzer executes code in a way it should not. 

> I'd welcome adding more fuzzers, indeed. Personally, I'd have no issue with having a single PR covering all of those, but I'm but a single maintainer. I can bring it up with other folks at the next call. In the meantime, can you please rebase so the GHA are run?

Great, we will get these uploaded when the whole OSS-Fuzz integraion has been completed. We can then migrate the fuzzers from https://github.com/cncf/cncf-fuzzing to here 

Notice that I did join the meeting around the time of this PR to discuss this work. Essentially it's work we have been doing with the Linux Foundation to support various CNCF projects

**Comment by AdamKorcz** [comment]
While we are waiting for review, which email address can we add for https://github.com/google/oss-fuzz/pull/6014 ?

**Comment by milosgajdos** [comment]
> While we are waiting for review, which email address can we add for [google/oss-fuzz#6014](https://github.com/google/oss-fuzz/pull/6014) ?

I dont think we have any specific email address to reach all maintainers of distribution, so unfortunately I've no answer for you here. I wonder if @SteveLasker might know?

**Comment by DavidKorczynski** [comment]
We can always add more later, i.e. it's fine to start with a single email address

**Review by wy65701436 — APPROVED** [review]
lgtm

**Comment by milosgajdos** [comment]
@AdamKorcz you should be able to email cncf-distribution-security at lists.cncf.io

**Comment by DavidKorczynski** [comment]
Thanks @milosgajdos - using a list won't work unfortunately. It needs to be email accounts linked to Google accounts as you need to log in to watch the reports. Example project.yamls for CNCF projects:
- [CNCF](https://github.com/google/oss-fuzz/blob/master/projects/envoy/project.yaml)
- [Fluent-bit](https://github.com/google/oss-fuzz/blob/master/projects/fluent-bit/project.yaml)
- [linkerd2](https://github.com/google/oss-fuzz/blob/master/projects/linkerd2/project.yaml)

**Comment by milosgajdos** [comment]
Well, in that case I'd recommend grabbing the list of email addresses from https://github.com/distribution/distribution/blob/main/MAINTAINERS

Do we need to do anything extra to get this sorted? Would apreciate some guidance here if the above is not enough.

**Comment by DavidKorczynski** [comment]
Thanks Milos! I will put those on.

Nothing more now, we have everything in terms of getting the fuzzers up and running. Will ping you on the PR that does the OSS-Fuzz integration on the OSS-Fuzz repository.

**Pull request merged**
